File: C++/examples/C++/SquareRotation.cpp
// main.cpp
#include <cmath>
#include <cstdlib>
#include <iostream>
#include <vector>
#include <chrono>

// --- GLUT portátil ---
#if defined(__APPLE__)
  #include <GLUT/glut.h>
#elif defined(_WIN32)
  #include <windows.h>
  #include <GL/freeglut.h>
#else
  #include <GL/glut.h>
#endif

// -----------------------------
// Estruturas de apoio
// -----------------------------
struct Ponto {
    float x{0.0f}, y{0.0f};
    Ponto() = default;
    Ponto(float x, float y) : x(x), y(y) {}

    void print() const { std::cout << "Ponto (" << x << ", " << y << ")\n"; }
    void set(float nx, float ny) { x = nx; y = ny; }

    Ponto operator+(const Ponto& o) const { return {x + o.x, y + o.y}; }
    Ponto operator-(const Ponto& o) const { return {x - o.x, y - o.y}; }
    Ponto operator*(float esc) const      { return {x * esc,   y * esc}; }
};

struct Quadrado {
    Ponto pos{0.f, 0.f};   // canto inferior esquerdo
    float w{0.25f};        // largura
    float h{0.25f};        // altura
    float r{0.f};          // rotação em graus
    float cr{1.f}, cg{0.f}, cb{0.f}; // cor

    Quadrado(float w, float h) : w(w), h(h) {}
    Quadrado(float w, float h, float cr, float cg, float cb)
        : w(w), h(h), cr(cr), cg(cg), cb(cb) {}
};

// -----------------------------
// Estado global (como no Python)
// -----------------------------
std::vector<Quadrado> quadrados = { Quadrado(0.25f, 0.25f) };
int num_quadrado = 0;

// câmera (ortho + pan)
float leftV = 0.f, rightV = 0.f, topV = 0.f, bottomV = 0.f;
float panX = 0.f, panY = 0.f;

// tempo/animacao (~30 FPS)
auto  t_last   = std::chrono::steady_clock::now();
double soma_dt = 0.0;

// -----------------------------
// Desenho dos eixos
// -----------------------------
void desenhaEixos() {
    glPushMatrix();
    glLoadIdentity();

    glColor3f(1.f, 1.f, 1.f);
    glLineWidth(1.f);

    glBegin(GL_LINES);
    glVertex2f(leftV,  0.f);
    glVertex2f(rightV, 0.f);
    glVertex2f(0.f, bottomV);
    glVertex2f(0.f, topV);
    glEnd();

    glPopMatrix();
}

// -----------------------------
// Desenho do quadrado (pos + rot)
// -----------------------------
void desenhaQuadrado(float x, float y, float w, float h, float r) {
    glPushMatrix();

    glTranslatef(x, y, 0.f);      // posição
    glRotatef(r, 0.f, 0.f, 1.f);  // rotação no plano XY

    glBegin(GL_QUADS);
    glVertex2f(0.f, 0.f);
    glVertex2f(w,  0.f);
    glVertex2f(w,  h);
    glVertex2f(0.f, h);
    glEnd();

    glPopMatrix();
}

// -----------------------------
// Display
// -----------------------------
void Desenha() {
    // Projeção ortográfica com pan
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluOrtho2D(leftV + panX, rightV + panX, bottomV + panY, topV + panY);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();

    // Fundo
    glClearColor(0.f, 0.f, 0.f, 1.f);
    glClear(GL_COLOR_BUFFER_BIT);

    // Desenha todos os quadrados
    for (const auto& q : quadrados) {
        glColor3f(q.cr, q.cg, q.cb);
        // Nota: mantendo semântica (w, h).
        desenhaQuadrado(q.pos.x, q.pos.y, q.w, q.h, q.r);
    }

    // Eixos
    desenhaEixos();

    glFlush(); // GLUT_SINGLE
}

// -----------------------------
// Idle / Animação (~30 FPS)
// -----------------------------
void Animacao() {
    auto t_now = std::chrono::steady_clock::now();
    std::chrono::duration<double> dt = t_now - t_last;
    t_last = t_now;

    soma_dt += dt.count();
    if (soma_dt > (1.0 / 30.0)) {
        soma_dt = 0.0;

        // Atualiza rotação do quadrado ativo (cíclica 0..359)
        Quadrado& q = quadrados[num_quadrado];
        q.r += 1.f;
        if (q.r >= 360.f) q.r -= 360.f;

        glutPostRedisplay();
    }
}

// -----------------------------
// Teclado ASCII (ESC, espaço, WASD)
// -----------------------------
void Teclado(unsigned char key, int, int) {
    if (key == 27) std::exit(0); // ESC

    if (key == ' ') {
        // cria novo quadrado e seleciona como atual
        quadrados.emplace_back(0.25f, 0.25f);
        num_quadrado = static_cast<int>(quadrados.size()) - 1;

        // esquema de cores como no Python:
        // r = ((n/9) % 3) / 2.0
        // g = ((n/3) % 3) / 2.0
        // b = (n % 3) / 2.0
        int n = num_quadrado;
        quadrados[num_quadrado].cr = std::fmod(n / 9.0f, 3.0f) / 2.0f;
        quadrados[num_quadrado].cg = std::fmod(n / 3.0f, 3.0f) / 2.0f;
        quadrados[num_quadrado].cb = (n % 3) / 2.0f;
    }

    const float step = 0.01f; // pan
    if (key == 'a' || key == 'A') panX -= step;
    if (key == 'd' || key == 'D') panX += step;
    if (key == 'w' || key == 'W') panY += step;
    if (key == 's' || key == 'S') panY -= step;

    glutPostRedisplay();
}

// -----------------------------
// Teclas especiais (setas) — move quadrado ativo
// -----------------------------
void TeclasEspeciais(int key, int, int) {
    const float dx = 0.02f, dy = 0.02f;

    if (key == GLUT_KEY_LEFT)  quadrados[num_quadrado].pos = quadrados[num_quadrado].pos - Ponto(dx, 0.f);
    if (key == GLUT_KEY_RIGHT) quadrados[num_quadrado].pos = quadrados[num_quadrado].pos + Ponto(dx, 0.f);
    if (key == GLUT_KEY_UP)    quadrados[num_quadrado].pos = quadrados[num_quadrado].pos + Ponto(0.f, dy);
    if (key == GLUT_KEY_DOWN)  quadrados[num_quadrado].pos = quadrados[num_quadrado].pos - Ponto(0.f, dy);

    glutPostRedisplay();
}

// -----------------------------
// Inicialização
// -----------------------------
void Inicializa() {
    glMatrixMode(GL_PROJECTION);
    leftV = -1.f; rightV = 1.f; topV = 1.f; bottomV = -1.f;
    gluOrtho2D(leftV + panX, rightV + panX, bottomV + panY, topV + panY);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();

    // reset de temporização
    t_last   = std::chrono::steady_clock::now();
    soma_dt  = 0.0;
}

// -----------------------------
// main
// -----------------------------
int main(int argc, char** argv) {
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB);
    glutInitWindowSize(800, 800);
    glutCreateWindow("Desenha OpenGL (C++ - Rotacao de Quadrado)");

    glutDisplayFunc(Desenha);
    glutIdleFunc(Animacao);
    glutKeyboardFunc(Teclado);
    glutSpecialFunc(TeclasEspeciais);

    Inicializa();
    glutMainLoop();
    return 0;
}

File: examples/Python/SquareRotation.py
import math
import time

from OpenGL.GLUT import *
from OpenGL.GLU import *
from OpenGL.GL import *

# Classe que representa um ponto 2D, com operações úteis para movimentação
class Ponto:
    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y

    def print(self) -> None:
        print("Ponto (", self.x, ",", self.y, ")")

    def set(self, x: float, y: float) -> None:
        self.x = x
        self.y = y

    # Operações matemáticas entre pontos
    def __add__(self, other):
        return Ponto(self.x + other.x, self.y + other.y)

    def __sub__(self, other):
        return Ponto(self.x - other.x, self.y - other.y)

    def __mul__(self, escalar: float):
        return Ponto(self.x * escalar, self.y * escalar)

# Classe que representa um quadrado
class Quadrado:
    def __init__(self, w, h):
        self.pos = Ponto(0, 0)    # Posição do quadrado
        self.w = w                # Largura
        self.h = h                # Altura
        self.c = (1, 0, 0)        # Cor (vermelho por padrão)
        self.r = 0                # Ângulo de rotação em graus

# Lista de quadrados e controle de qual está ativo
quadrados = [Quadrado(0.25, 0.25)]
num_quadrado = 0

# Controle de tempo para animação
tempo_antes = time.time()
soma_dt = 0

# Parâmetros da câmera (pan e viewport)
left = 0
right = 0
top = 0
bottom = 0
panX = 0
panY = 0

# Desenha os eixos X e Y no mundo
def desenhaEixos():
    glPushMatrix()
    glLoadIdentity()

    glColor3f(1, 1, 1)
    glLineWidth(1)

    glBegin(GL_LINES)
    glVertex2f(left, 0)
    glVertex2f(right, 0)
    glVertex2f(0, bottom)
    glVertex2f(0, top)
    glEnd()

    glPopMatrix()

# Desenha um quadrado na posição (x, y), com rotação r
def desenhaQuadrado(x, y, w, h, r):
    glPushMatrix()

    glTranslatef(x, y, 0)     # Move para a posição desejada
    glRotatef(r, 0, 0, 1)     # Rotaciona em torno do eixo Z (plano XY)

    glBegin(GL_QUADS)
    glVertex2f(0, 0)
    glVertex2f(w, 0)
    glVertex2f(w, h)
    glVertex2f(0, h)
    glEnd()

    glPopMatrix()

# Função principal de redesenho da cena
def Desenha():
    global left, right, top, bottom, panX, panY

    # Define projeção ortográfica com pan
    glMatrixMode(GL_PROJECTION)
    glLoadIdentity()
    gluOrtho2D(left + panX, right + panX, bottom + panY, top + panY)
    glMatrixMode(GL_MODELVIEW)

    # Limpa a tela
    glClearColor(0, 0, 0, 1)
    glClear(GL_COLOR_BUFFER_BIT)

    # Desenha todos os quadrados
    for quadrado in quadrados:
        glColor3f(*quadrado.c)
        desenhaQuadrado(quadrado.pos.x, quadrado.pos.y, quadrado.h, quadrado.w, quadrado.r)

    # Desenha os eixos (ajuda visual)
    desenhaEixos()

    # Finaliza os comandos de renderização
    glFlush()

# Função chamada constantemente (idle) para atualizar a animação
def Animacao():
    global soma_dt, tempo_antes

    tempo_agora = time.time()
    delta_time = tempo_agora - tempo_antes
    tempo_antes = tempo_agora

    soma_dt += delta_time

    if soma_dt > 1.0 / 30:  # Aproximadamente 30 quadros por segundo
        soma_dt = 0

        # Atualiza o ângulo de rotação do quadrado atual
        # O % 360 garante que o valor fique entre 0 e 359 (rotação cíclica)
        quadrados[num_quadrado].r = (quadrados[num_quadrado].r + 1) % 360

        glutPostRedisplay()  # Solicita redesenho da tela

# Gerencia as teclas do teclado comum (ASCII)
def Teclado(key: chr, x: int, y: int):
    global num_quadrado, panX, panY

    if key == 27:  # ESC
        exit(0)

    if key == b' ':
        # Cria um novo quadrado e incrementa o índice
        quadrados.append(Quadrado(0.25, 0.25))
        num_quadrado += 1
        quadrados[num_quadrado].c = (
            num_quadrado / 9 % 3 / 2.0,
            num_quadrado / 3 % 3 / 2.0,
            num_quadrado % 3 / 2.0
        )

    # Controle da câmera (pan) com WASD
    if key == b'a':
        panX -= 0.01
    if key == b'd':
        panX += 0.01
    if key == b'w':
        panY += 0.01
    if key == b's':
        panY -= 0.01

    glutPostRedisplay()

# Gerencia teclas especiais (setas) para mover o quadrado ativo
def TeclasEspeciais(key: int, x: int, y: int):
    if key == GLUT_KEY_LEFT:
        quadrados[num_quadrado].pos -= Ponto(0.02, 0)
    if key == GLUT_KEY_RIGHT:
        quadrados[num_quadrado].pos += Ponto(0.02, 0)
    if key == GLUT_KEY_UP:
        quadrados[num_quadrado].pos += Ponto(0, 0.02)
    if key == GLUT_KEY_DOWN:
        quadrados[num_quadrado].pos -= Ponto(0, 0.02)

    glutPostRedisplay()

# Inicializa os parâmetros da câmera
def Inicializa():
    global left, right, top, bottom, panX, panY

    glMatrixMode(GL_PROJECTION)
    left = -1
    right = 1
    top = 1
    bottom = -1
    gluOrtho2D(left + panX, right + panX, bottom + panY, top + panY)
    glMatrixMode(GL_MODELVIEW)

# Função principal
def main():
    glutInit(sys.argv)
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB)
    glutInitWindowSize(800, 800)
    glutCreateWindow(b"Desenha OpenGL")

    glutDisplayFunc(Desenha)
    glutIdleFunc(Animacao)
    glutKeyboardFunc(Teclado)
    glutSpecialFunc(TeclasEspeciais)

    Inicializa()

    try:
        glutMainLoop()
    except SystemExit:
        pass

if __name__ == '__main__':
    main()